File: Naumen/Scala/app/services/ShipsFactoryService.scala
package services

import javax.inject._
import models._

@Singleton
class ShipsFactoryService @Inject() () {
  private var _shipsFactory: ShipsFactory = null

  def setupNewFactory (size: Int): Unit = {
    _shipsFactory = new ShipsFactory(size)
  }

  def addShipInQueue(ship: Ship): Unit = {
    if (_shipsFactory == null) {
      throw new NullPointerException("Factory should be setup before use.")
    }
    _shipsFactory.addShipInQueue(ship)
  }

    def nextRepairStartTime(): RepairTime = {
      if (_shipsFactory == null)
        null
      else
        _shipsFactory.nextRepairStartTime()
    }
}

File: Naumen/CSharp/NaumenCSharp/Services/ShipsFactoryService.cs
using System;
using NaumenCSharp.Models;

namespace NaumenCSharp.Services
{
    public class ShipsFactoryService
    {
        private ShipsFactory _shipsFactory;

        public void SetupNewFactory(int size)
        {
            _shipsFactory = new ShipsFactory(size);
        }

        public void AddShipInQueue(Ship ship)
        {
            CheckFactorySetup();
            _shipsFactory.AddShipInQueue(ship);
        }

        private void CheckFactorySetup()
        {
            if (_shipsFactory == null)
                throw new ArgumentNullException(nameof(_shipsFactory), "Factory should be setup before use.");
        }

        public int? GetNextRepairStartTime()
        {
            return _shipsFactory?.GetNextShipRepairStart();
        }
    }
}